{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 1. Combine all the sequences to one file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat genbank_sequences/*.fasta > demo_data/all_seqs.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Check the number of sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of sequences in combined file: 39\n"
     ]
    }
   ],
   "source": [
    "with open(\"demo_data/all_seqs.fa\", \"r\") as fasta_file:\n",
    "    content = fasta_file.read()\n",
    "    num_sequences = content.count(\">\")\n",
    "print(f\"The number of sequences in combined file: {num_sequences}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 2. Multiple Sequence Alignment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 4 ambiguous characters.\n",
      "    1 / 39\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "   30 / 39\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP    24 / 38 \n",
      "Reallocating..done. *alloclen = 14286\n",
      "STEP    38 / 38 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "    0 / 39\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "   30 / 39\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP    35 / 38 \n",
      "Reallocating..done. *alloclen = 14282\n",
      "STEP    38 / 38 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.505\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "dndpre (nuc) Version 7.505\n",
      "alg=X, model=DNA200 (2), 1.53 (4.59), 0.37 (1.11), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "   30 / 39\n",
      "Segment   1/183    1- 177\n",
      "Segment   2/183  172- 202.    identical.    rejected. identical.   \n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   3/183  202- 299\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   4/183  299- 358\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   5/183  358- 399\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   6/183  399- 410\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   7/183  410- 564\n",
      "STEP 002-007-1  identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   8/183  554- 562\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   9/183  562- 570\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  10/183  570- 583\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  11/183  583- 614\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  12/183  614- 617\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  13/183  617- 656\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  14/183  656- 704\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  15/183  704- 935\n",
      "STEP 002-037-1  identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  16/183  935- 939\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  17/183  939- 941\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  18/183  941- 945\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  19/183  945-1509\n",
      "Segment  20/183 1503-1505.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    rejected. accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  21/183 1505-1512\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  22/183 1512-1516\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  23/183 1516-1526\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  24/183 1526-1529\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  25/183 1529-1533\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  26/183 1533-1574\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  27/183 1574-1583\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  28/183 1583-1639\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  29/183 1639-1778\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  30/183 1778-1788\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  31/183 1788-1790\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  32/183 1790-1813\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  33/183 1813-1820\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  34/183 1820-1824\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  35/183 1824-1857\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  36/183 1857-1938\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  37/183 1938-1959\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  38/183 1959-2119\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  39/183 2119-2135\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  40/183 2135-2141\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  41/183 2141-2144\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  42/183 2144-2170\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  43/183 2170-2183\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  44/183 2183-2191\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  45/183 2191-2198\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  46/183 2198-2230\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  47/183 2230-2240\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  48/183 2240-2246\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  49/183 2246-2269\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  50/183 2269-2298\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  51/183 2298-2300\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  52/183 2300-2323\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  53/183 2323-2447\n",
      "Segment  54/183 2447-2456.    accepted. identical.   \n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  55/183 2456-2459\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  56/183 2459-2537\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  57/183 2537-2567\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  58/183 2567-2606\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  59/183 2606-2614\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  60/183 2614-2642\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  61/183 2642-2645\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  62/183 2645-2651\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  63/183 2651-2758\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  64/183 2758-2824\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  65/183 2824-2840\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  66/183 2840-2868\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  67/183 2868-2923\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  68/183 2923-3018\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  69/183 3018-3110\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  70/183 3110-3119\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  71/183 3119-3123\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  72/183 3123-3131\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  73/183 3131-3143\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  74/183 3143-3282\n",
      "STEP 002-007-0  identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  75/183 3282-3330\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  76/183 3330-3393\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  77/183 3393-3441\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  78/183 3441-3451\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  79/183 3451-3455\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  80/183 3455-3464\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  81/183 3464-3467\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  82/183 3467-3471\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  83/183 3471-3473\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  84/183 3473-3525\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  85/183 3525-3547\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  86/183 3547-3551\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  87/183 3551-3560\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  88/183 3560-3605\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  89/183 3605-3652\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  90/183 3652-3657\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  91/183 3657-3665\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  92/183 3665-3697\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  93/183 3697-3738\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  94/183 3738-3758\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  95/183 3758-3773\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  96/183 3773-3806\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  97/183 3806-3814\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  98/183 3814-3824\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  99/183 3824-3829\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 100/183 3829-3833\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 101/183 3833-3890\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 102/183 3890-3938\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 103/183 3938-3956\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 104/183 3956-4034\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 105/183 4034-4044\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 106/183 4044-4074\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 107/183 4074-4082\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 108/183 4082-4088\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 109/183 4088-4095\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 110/183 4095-4097\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 111/183 4097-4100\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 112/183 4100-4105\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 113/183 4105-4134\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 114/183 4134-4215\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 115/183 4215-4247\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 116/183 4247-4257\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 117/183 4257-4288\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 118/183 4288-4319\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 119/183 4319-4338\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 120/183 4338-4381\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 121/183 4381-4385\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 122/183 4385-4454\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 123/183 4454-4495\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 124/183 4495-4572\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 125/183 4572-4585\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 126/183 4585-4595\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 127/183 4595-4601\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 128/183 4601-4634\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 129/183 4634-4742\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 130/183 4742-4781\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 131/183 4781-4794\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 132/183 4794-4843\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 133/183 4843-4871\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 134/183 4871-4931\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 135/183 4931-4934\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 136/183 4934-5014\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 137/183 5014-5026\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 138/183 5026-5076\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 139/183 5076-5078\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 140/183 5078-5143\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 141/183 5143-5151\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 142/183 5151-5263\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 143/183 5263-5341\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 144/183 5341-5366\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 145/183 5366-5372\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 146/183 5372-5378\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 147/183 5378-5401\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 148/183 5401-5413\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 149/183 5413-5423\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 150/183 5423-5432\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 151/183 5432-5436\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 152/183 5436-5450\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 153/183 5450-5515\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 154/183 5515-5547\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 155/183 5547-5636\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 156/183 5636-5677\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 157/183 5677-5683\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 158/183 5683-5687\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 159/183 5687-5696\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 160/183 5696-5709\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 161/183 5709-5771\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 162/183 5771-5774\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 163/183 5774-5810\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 164/183 5810-5829\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 165/183 5829-5840\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 166/183 5840-5875\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 167/183 5875-6109\n",
      "Segment 168/183 6109-6116.    identical.    identical.    identical.    identical.   \n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 169/183 6116-6120\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 170/183 6120-6153\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 171/183 6153-6165\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 172/183 6165-6254\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 173/183 6254-6266\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 174/183 6266-6298\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 175/183 6298-6302\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 176/183 6302-6326\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 177/183 6326-6332\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 178/183 6332-6336\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 179/183 6336-6408\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 180/183 6408-6514\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 181/183 6514-6529\n",
      "STEP 002-037-1  identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 182/183 6529-6536\n",
      "STEP 002-037-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment 183/183 6536-6772\n",
      "done 002-001-1  identical.    identical.    identical.    accepted. identical.   \n",
      "dvtditr (nuc) Version 7.505\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-i (Standard)\n",
      " Iterative refinement method (max. 2 iterations)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! mafft --auto demo_data/all_seqs.fa > demo_data/all_seqs_mafft.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 3. Launching ModelFinder to find the best substitution model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    IliaPC (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree2 -m MFP -s demo_data/all_seqs_mafft.fa --prefix demo_data/tree_MF2 -T AUTO\n",
      "Seed:    524089 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Thu Apr 25 23:16:32 2024\n",
      "Kernel:  AVX+FMA - auto-detect threads (24 CPU cores detected)\n",
      "\n",
      "Reading alignment file demo_data/all_seqs_mafft.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 39 sequences with 6771 columns, 4754 distinct patterns\n",
      "4433 parsimony-informative, 567 singleton sites, 1771 constant sites\n",
      "             Gap/Ambiguity  Composition  p-value\n",
      "   1  FJ593498.1     3.07%    failed      0.00%\n",
      "   2  KX512433.1     3.07%    failed      0.01%\n",
      "   3  KX779126.1     4.36%    failed      2.13%\n",
      "   4  LC553715.1     3.57%    passed     43.91%\n",
      "   5  MG663536.1     3.49%    failed      4.98%\n",
      "   6  NC_003468.2    3.09%    passed     91.21%\n",
      "   7  NC_005222.1    3.51%    passed      8.12%\n",
      "   8  NC_005225.1    3.26%    passed     17.84%\n",
      "   9  NC_005235.1    3.53%    failed      2.24%\n",
      "  10  NC_005238.1    3.56%    passed     29.17%\n",
      "  11  NC_006435.1    3.51%    passed     21.03%\n",
      "  12  NC_010707.1    2.81%    passed     44.32%\n",
      "  13  NC_034399.1    2.70%    passed     10.59%\n",
      "  14  NC_034401.1    3.18%    failed      1.00%\n",
      "  15  NC_034402.1    3.44%    failed      0.00%\n",
      "  16  NC_034403.1    3.57%    passed      5.09%\n",
      "  17  NC_034407.1    3.25%    passed     18.98%\n",
      "  18  NC_034467.1    3.54%    passed     10.36%\n",
      "  19  NC_034485.1    3.51%    passed     48.84%\n",
      "  20  NC_034515.1    2.85%    passed     12.60%\n",
      "  21  NC_034517.1    3.54%    failed      1.80%\n",
      "  22  NC_034519.1    2.81%    passed     71.75%\n",
      "  23  NC_034553.1    3.07%    passed     97.88%\n",
      "  24  NC_034556.1    3.88%    passed     37.56%\n",
      "  25  NC_034560.1    4.02%    passed     70.93%\n",
      "  26  NC_034564.1    2.81%    failed      4.62%\n",
      "  27  NC_038299.1    3.20%    passed      6.20%\n",
      "  28  NC_038515.1    3.54%    passed     31.01%\n",
      "  29  NC_038529.1    3.46%    failed      2.66%\n",
      "  30  NC_038695.1    3.15%    failed      0.01%\n",
      "  31  NC_038939.1    3.13%    failed      0.10%\n",
      "  32  NC_055147.1    3.53%    passed     39.02%\n",
      "  33  NC_055170.1    2.13%    failed      1.72%\n",
      "  34  NC_055632.1    4.79%    failed      0.06%\n",
      "  35  NC_055636.1    4.52%    passed     30.57%\n",
      "  36  NC_077671.1    3.09%    passed     19.53%\n",
      "  37  NC_078262.1    3.68%    passed     12.58%\n",
      "  38  NC_078485.1    3.54%    passed      7.42%\n",
      "  39  OR684449.1     7.09%    failed      2.04%\n",
      "****  TOTAL          3.48%  15 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.028 seconds\n",
      "Measuring multi-threading efficiency up to 24 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "56 trees examined\n",
      "Threads: 1 / Time: 24.131 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -192853\n",
      "Threads: 2 / Time: 13.931 sec / Speedup: 1.732 / Efficiency: 87% / LogL: -192853\n",
      "Threads: 3 / Time: 10.659 sec / Speedup: 2.264 / Efficiency: 75% / LogL: -192853\n",
      "Threads: 4 / Time: 11.198 sec / Speedup: 2.155 / Efficiency: 54% / LogL: -192853\n",
      "Threads: 5 / Time: 7.503 sec / Speedup: 3.216 / Efficiency: 64% / LogL: -192853\n",
      "Threads: 6 / Time: 5.659 sec / Speedup: 4.264 / Efficiency: 71% / LogL: -192853\n",
      "Threads: 7 / Time: 5.465 sec / Speedup: 4.415 / Efficiency: 63% / LogL: -192853\n",
      "Threads: 8 / Time: 5.292 sec / Speedup: 4.560 / Efficiency: 57% / LogL: -192853\n",
      "Threads: 9 / Time: 5.483 sec / Speedup: 4.401 / Efficiency: 49% / LogL: -192853\n",
      "BEST NUMBER OF THREADS: 8\n",
      "\n",
      "Perform fast likelihood tree search using GTR+I+G model...\n",
      "Estimate model parameters (epsilon = 5.000)\n",
      "Perform nearest neighbor interchange...\n",
      "Estimate model parameters (epsilon = 1.000)\n",
      "1. Initial log-likelihood: -169331.512\n",
      "2. Current log-likelihood: -169329.262\n",
      "3. Current log-likelihood: -169328.095\n",
      "Optimal log-likelihood: -169326.968\n",
      "Rate parameters:  A-C: 7.23424  A-G: 9.10133  A-T: 2.38571  C-G: 3.68238  C-T: 20.46609  G-T: 1.00000\n",
      "Base frequencies:  A: 0.330  C: 0.156  G: 0.211  T: 0.303\n",
      "Proportion of invariable sites: 0.192\n",
      "Gamma shape alpha: 0.602\n",
      "Parameters optimization took 3 rounds (0.228 sec)\n",
      "Time for fast ML tree search: 1.494 seconds\n",
      "\n",
      "NOTE: ModelFinder requires 60 MB RAM!\n",
      "ModelFinder will test up to 286 DNA models (sample size: 6771) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  GTR+F         192561.916   83  385289.831   385291.916   385855.925\n",
      "  2  GTR+F+I       180323.899   84  360815.798   360817.934   361388.712\n",
      "  3  GTR+F+G4      169570.784   84  339309.569   339311.705   339882.483\n",
      "  4  GTR+F+I+G4    169324.853   85  338819.706   338821.893   339399.441\n",
      "  5  GTR+F+R2      173013.068   85  346196.136   346198.323   346775.870\n",
      "  6  GTR+F+R3      170054.933   87  340283.866   340286.157   340877.241\n",
      "  7  GTR+F+R4      169379.627   89  338937.254   338939.651   339544.269\n",
      "  8  GTR+F+R5      169158.046   91  338498.092   338500.598   339118.748\n",
      "  9  GTR+F+R6      169139.691   93  338465.383   338468.001   339099.680\n",
      " 10  GTR+F+R7      169124.138   95  338438.276   338441.008   339086.214\n",
      " 11  GTR+F+R8      169123.859   97  338441.718   338444.567   339103.297\n",
      " 23  SYM+R7        171914.024   92  344012.048   344014.610   344639.525\n",
      " 36  TVM+F+R7      169596.420   94  339380.840   339383.516   340021.958\n",
      " 49  TVMe+R7       171928.067   91  344038.135   344040.641   344658.791\n",
      " 62  TIM3+F+R7     169399.228   93  338984.456   338987.074   339618.753\n",
      " 75  TIM3e+R7      173212.066   90  346604.132   346606.584   347217.968\n",
      " 88  TIM2+F+R7     169711.891   93  339609.781   339612.400   340244.079\n",
      "101  TIM2e+R7      172243.948   90  344667.896   344670.348   345281.732\n",
      "114  TIM+F+R7      170011.566   93  340209.133   340211.751   340843.430\n",
      "127  TIMe+R7       172866.865   90  345913.731   345916.183   346527.567\n",
      "140  TPM3u+F+R7    169786.425   92  339756.850   339759.412   340384.327\n",
      "153  TPM3+F+R7     169786.425   92  339756.850   339759.412   340384.327\n",
      "166  TPM2u+F+R7    170047.710   92  340279.420   340281.982   340906.897\n",
      "179  TPM2+F+R7     170047.617   92  340279.233   340281.796   340906.710\n",
      "192  K3Pu+F+R7     170277.907   92  340739.814   340742.376   341367.291\n",
      "205  K3P+R7        172892.366   89  345962.731   345965.129   346569.747\n",
      "218  TN+F+R7       170011.762   92  340207.524   340210.086   340835.001\n",
      "231  TNe+R7        173310.580   89  346799.161   346801.559   347406.177\n",
      "244  HKY+F+R7      170280.457   91  340742.914   340745.421   341363.571\n",
      "257  K2P+R7        173341.823   88  346859.647   346861.991   347459.842\n",
      "270  F81+F+R7      175058.561   90  350297.122   350299.574   350910.958\n",
      "283  JC+R7         176385.636   87  352945.272   352947.563   353538.647\n",
      "Akaike Information Criterion:           GTR+F+R7\n",
      "Corrected Akaike Information Criterion: GTR+F+R7\n",
      "Bayesian Information Criterion:         GTR+F+R7\n",
      "Best-fit model: GTR+F+R7 chosen according to BIC\n",
      "\n",
      "All model information printed to demo_data/tree_MF2.model.gz\n",
      "CPU time for ModelFinder: 1616.847 seconds (0h:26m:56s)\n",
      "Wall-clock time for ModelFinder: 250.497 seconds (0h:4m:10s)\n",
      "\n",
      "NOTE: 42 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -169124.138\n",
      "2. Current log-likelihood: -169123.569\n",
      "3. Current log-likelihood: -169123.339\n",
      "4. Current log-likelihood: -169123.232\n",
      "Optimal log-likelihood: -169123.148\n",
      "Rate parameters:  A-C: 8.05494  A-G: 9.97402  A-T: 2.38683  C-G: 4.11119  C-T: 23.65448  G-T: 1.00000\n",
      "Base frequencies:  A: 0.330  C: 0.156  G: 0.211  T: 0.303\n",
      "Site proportion and rates:  (0.221,0.004) (0.131,0.047) (0.136,0.189) (0.169,0.529) (0.114,0.939) (0.136,1.903) (0.093,5.505)\n",
      "Parameters optimization took 4 rounds (1.247 sec)\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.021370 sec (of wall-clock time) 0.117299 sec(of CPU time)\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing RapidNJ tree took 0.081544 sec (of wall-clock time) 0.029135 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -169165.187\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 3.889 second\n",
      "Computing log-likelihood of 98 initial trees ... 2.177 seconds\n",
      "Current best score: -169123.148\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -169123.052\n",
      "Iteration 10 / LogL: -169188.520 / Time: 0h:0m:11s\n",
      "Iteration 20 / LogL: -169151.480 / Time: 0h:0m:15s\n",
      "Finish initializing candidate tree set (2)\n",
      "Current best tree score: -169123.052 / CPU time: 13.678\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -169129.457 / Time: 0h:0m:18s (0h:0m:46s left)\n",
      "Iteration 40 / LogL: -169123.529 / Time: 0h:0m:22s (0h:0m:35s left)\n",
      "Iteration 50 / LogL: -169127.024 / Time: 0h:0m:26s (0h:0m:27s left)\n",
      "Iteration 60 / LogL: -169124.588 / Time: 0h:0m:29s (0h:0m:20s left)\n",
      "Iteration 70 / LogL: -169123.231 / Time: 0h:0m:33s (0h:0m:14s left)\n",
      "Iteration 80 / LogL: -169124.343 / Time: 0h:0m:36s (0h:0m:9s left)\n",
      "Iteration 90 / LogL: -169125.319 / Time: 0h:0m:40s (0h:0m:5s left)\n",
      "Iteration 100 / LogL: -169123.948 / Time: 0h:0m:44s (0h:0m:0s left)\n",
      "TREE SEARCH COMPLETED AFTER 102 ITERATIONS / Time: 0h:0m:45s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -169123.052\n",
      "2. Current log-likelihood: -169122.990\n",
      "3. Current log-likelihood: -169122.943\n",
      "4. Current log-likelihood: -169122.900\n",
      "5. Current log-likelihood: -169122.862\n",
      "6. Current log-likelihood: -169122.833\n",
      "7. Current log-likelihood: -169122.807\n",
      "8. Current log-likelihood: -169122.775\n",
      "9. Current log-likelihood: -169122.753\n",
      "Optimal log-likelihood: -169122.740\n",
      "Rate parameters:  A-C: 8.07787  A-G: 10.02072  A-T: 2.39763  C-G: 4.12799  C-T: 23.68921  G-T: 1.00000\n",
      "Base frequencies:  A: 0.330  C: 0.156  G: 0.211  T: 0.303\n",
      "Site proportion and rates:  (0.213,0.004) (0.137,0.044) (0.138,0.187) (0.174,0.531) (0.114,0.966) (0.133,1.935) (0.091,5.563)\n",
      "Parameters optimization took 9 rounds (2.614 sec)\n",
      "BEST SCORE FOUND : -169122.740\n",
      "Total tree length: 23.965\n",
      "\n",
      "Total number of iterations: 102\n",
      "CPU time used for tree search: 320.655 sec (0h:5m:20s)\n",
      "Wall-clock time used for tree search: 43.569 sec (0h:0m:43s)\n",
      "Total CPU time used: 354.654 sec (0h:5m:54s)\n",
      "Total wall-clock time used: 47.973 sec (0h:0m:47s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                demo_data/tree_MF2.iqtree\n",
      "  Maximum-likelihood tree:       demo_data/tree_MF2.treefile\n",
      "  Likelihood distances:          demo_data/tree_MF2.mldist\n",
      "  Screen log file:               demo_data/tree_MF2.log\n",
      "\n",
      "Date and Time: Thu Apr 25 23:21:31 2024\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -m MFP -s demo_data/all_seqs_mafft.fa --prefix demo_data/tree_MF2 -T AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Getting the best substitution model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best-fit model according to BIC: GTR+F+R7\n",
      "\n",
      "List of models sorted by BIC scores: \n",
      "\n",
      "Model                  LogL         AIC      w-AIC        AICc     w-AICc         BIC      w-BIC\n",
      "GTR+F+R7        -169124.138  338438.276 +    0.848  338441.008 +    0.856  339086.214 +    0.999\n"
     ]
    }
   ],
   "source": [
    "! head -42 demo_data/tree_MF2.iqtree | tail -6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 4. Launching IQ-TREE to construct the tree**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    IliaPC (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree2 -s demo_data/all_seqs_mafft.fa -m GTR+F+R7 -pre demo_data/tree_ufb -bb 1000 -nt AUTO\n",
      "Seed:    622575 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Thu Apr 25 23:21:44 2024\n",
      "Kernel:  AVX+FMA - auto-detect threads (24 CPU cores detected)\n",
      "\n",
      "Reading alignment file demo_data/all_seqs_mafft.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 39 sequences with 6771 columns, 4754 distinct patterns\n",
      "4433 parsimony-informative, 567 singleton sites, 1771 constant sites\n",
      "             Gap/Ambiguity  Composition  p-value\n",
      "   1  FJ593498.1     3.07%    failed      0.00%\n",
      "   2  KX512433.1     3.07%    failed      0.01%\n",
      "   3  KX779126.1     4.36%    failed      2.13%\n",
      "   4  LC553715.1     3.57%    passed     43.91%\n",
      "   5  MG663536.1     3.49%    failed      4.98%\n",
      "   6  NC_003468.2    3.09%    passed     91.21%\n",
      "   7  NC_005222.1    3.51%    passed      8.12%\n",
      "   8  NC_005225.1    3.26%    passed     17.84%\n",
      "   9  NC_005235.1    3.53%    failed      2.24%\n",
      "  10  NC_005238.1    3.56%    passed     29.17%\n",
      "  11  NC_006435.1    3.51%    passed     21.03%\n",
      "  12  NC_010707.1    2.81%    passed     44.32%\n",
      "  13  NC_034399.1    2.70%    passed     10.59%\n",
      "  14  NC_034401.1    3.18%    failed      1.00%\n",
      "  15  NC_034402.1    3.44%    failed      0.00%\n",
      "  16  NC_034403.1    3.57%    passed      5.09%\n",
      "  17  NC_034407.1    3.25%    passed     18.98%\n",
      "  18  NC_034467.1    3.54%    passed     10.36%\n",
      "  19  NC_034485.1    3.51%    passed     48.84%\n",
      "  20  NC_034515.1    2.85%    passed     12.60%\n",
      "  21  NC_034517.1    3.54%    failed      1.80%\n",
      "  22  NC_034519.1    2.81%    passed     71.75%\n",
      "  23  NC_034553.1    3.07%    passed     97.88%\n",
      "  24  NC_034556.1    3.88%    passed     37.56%\n",
      "  25  NC_034560.1    4.02%    passed     70.93%\n",
      "  26  NC_034564.1    2.81%    failed      4.62%\n",
      "  27  NC_038299.1    3.20%    passed      6.20%\n",
      "  28  NC_038515.1    3.54%    passed     31.01%\n",
      "  29  NC_038529.1    3.46%    failed      2.66%\n",
      "  30  NC_038695.1    3.15%    failed      0.01%\n",
      "  31  NC_038939.1    3.13%    failed      0.10%\n",
      "  32  NC_055147.1    3.53%    passed     39.02%\n",
      "  33  NC_055170.1    2.13%    failed      1.72%\n",
      "  34  NC_055632.1    4.79%    failed      0.06%\n",
      "  35  NC_055636.1    4.52%    passed     30.57%\n",
      "  36  NC_077671.1    3.09%    passed     19.53%\n",
      "  37  NC_078262.1    3.68%    passed     12.58%\n",
      "  38  NC_078485.1    3.54%    passed      7.42%\n",
      "  39  OR684449.1     7.09%    failed      2.04%\n",
      "****  TOTAL          3.48%  15 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.032 seconds\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 622575)...\n",
      "\n",
      "NOTE: 60 MB RAM (0 GB) is required!\n",
      "Measuring multi-threading efficiency up to 24 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "36 trees examined\n",
      "Threads: 1 / Time: 24.012 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -195740\n",
      "Threads: 2 / Time: 12.994 sec / Speedup: 1.848 / Efficiency: 92% / LogL: -195740\n",
      "Threads: 3 / Time: 9.675 sec / Speedup: 2.482 / Efficiency: 83% / LogL: -195740\n",
      "Threads: 4 / Time: 7.354 sec / Speedup: 3.265 / Efficiency: 82% / LogL: -195740\n",
      "Threads: 5 / Time: 6.416 sec / Speedup: 3.743 / Efficiency: 75% / LogL: -195740\n",
      "Threads: 6 / Time: 5.753 sec / Speedup: 4.174 / Efficiency: 70% / LogL: -195740\n",
      "Threads: 7 / Time: 5.850 sec / Speedup: 4.105 / Efficiency: 59% / LogL: -195740\n",
      "Threads: 8 / Time: 5.828 sec / Speedup: 4.120 / Efficiency: 51% / LogL: -195740\n",
      "Threads: 9 / Time: 5.518 sec / Speedup: 4.352 / Efficiency: 48% / LogL: -195740\n",
      "BEST NUMBER OF THREADS: 6\n",
      "\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "1. Initial log-likelihood: -177626.058\n",
      "2. Current log-likelihood: -169794.670\n",
      "3. Current log-likelihood: -169359.228\n",
      "4. Current log-likelihood: -169270.234\n",
      "5. Current log-likelihood: -169233.068\n",
      "6. Current log-likelihood: -169221.588\n",
      "7. Current log-likelihood: -169216.811\n",
      "8. Current log-likelihood: -169214.397\n",
      "9. Current log-likelihood: -169212.860\n",
      "10. Current log-likelihood: -169211.700\n",
      "11. Current log-likelihood: -169210.883\n",
      "12. Current log-likelihood: -169210.116\n",
      "13. Current log-likelihood: -169209.432\n",
      "14. Current log-likelihood: -169208.881\n",
      "15. Current log-likelihood: -169208.362\n",
      "16. Current log-likelihood: -169207.794\n",
      "17. Current log-likelihood: -169207.304\n",
      "18. Current log-likelihood: -169206.862\n",
      "19. Current log-likelihood: -169206.449\n",
      "20. Current log-likelihood: -169206.072\n",
      "21. Current log-likelihood: -169205.712\n",
      "22. Current log-likelihood: -169205.383\n",
      "23. Current log-likelihood: -169205.075\n",
      "24. Current log-likelihood: -169204.790\n",
      "25. Current log-likelihood: -169204.519\n",
      "26. Current log-likelihood: -169204.269\n",
      "27. Current log-likelihood: -169204.038\n",
      "28. Current log-likelihood: -169203.787\n",
      "29. Current log-likelihood: -169203.564\n",
      "30. Current log-likelihood: -169203.371\n",
      "31. Current log-likelihood: -169203.192\n",
      "32. Current log-likelihood: -169203.028\n",
      "33. Current log-likelihood: -169202.873\n",
      "34. Current log-likelihood: -169202.722\n",
      "35. Current log-likelihood: -169202.583\n",
      "36. Current log-likelihood: -169202.448\n",
      "37. Current log-likelihood: -169202.320\n",
      "38. Current log-likelihood: -169202.203\n",
      "39. Current log-likelihood: -169202.088\n",
      "40. Current log-likelihood: -169201.980\n",
      "41. Current log-likelihood: -169201.879\n",
      "Optimal log-likelihood: -169201.728\n",
      "Rate parameters:  A-C: 8.03179  A-G: 9.95264  A-T: 2.42924  C-G: 4.09018  C-T: 23.43913  G-T: 1.00000\n",
      "Base frequencies:  A: 0.330  C: 0.156  G: 0.211  T: 0.303\n",
      "Site proportion and rates:  (0.216,0.004) (0.135,0.047) (0.134,0.192) (0.149,0.514) (0.131,0.892) (0.137,1.889) (0.097,5.290)\n",
      "Parameters optimization took 41 rounds (16.206 sec)\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.016208 sec (of wall-clock time) 0.087238 sec(of CPU time)\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing RapidNJ tree took 0.056935 sec (of wall-clock time) 0.073189 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -169166.277\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 3.959 second\n",
      "Computing log-likelihood of 98 initial trees ... 2.497 seconds\n",
      "Current best score: -169166.277\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -169123.352\n",
      "Iteration 10 / LogL: -169152.683 / Time: 0h:1m:53s\n",
      "Iteration 20 / LogL: -169187.042 / Time: 0h:1m:59s\n",
      "Finish initializing candidate tree set (1)\n",
      "Current best tree score: -169123.352 / CPU time: 19.379\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -169123.852 / Time: 0h:2m:3s (0h:5m:2s left)\n",
      "Iteration 40 / LogL: -169126.756 / Time: 0h:2m:7s (0h:3m:18s left)\n",
      "Iteration 50 / LogL: -169125.203 / Time: 0h:2m:10s (0h:2m:16s left)\n",
      "Log-likelihood cutoff on original alignment: -169186.542\n",
      "Iteration 60 / LogL: -169128.228 / Time: 0h:2m:14s (0h:1m:33s left)\n",
      "Iteration 70 / LogL: -169123.426 / Time: 0h:2m:17s (0h:1m:1s left)\n",
      "Iteration 80 / LogL: -169126.442 / Time: 0h:2m:21s (0h:0m:37s left)\n",
      "Iteration 90 / LogL: -169123.504 / Time: 0h:2m:24s (0h:0m:17s left)\n",
      "Iteration 100 / LogL: -169124.086 / Time: 0h:2m:28s (0h:0m:1s left)\n",
      "Log-likelihood cutoff on original alignment: -169186.964\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.993\n",
      "TREE SEARCH COMPLETED AFTER 102 ITERATIONS / Time: 0h:2m:29s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -169123.352\n",
      "2. Current log-likelihood: -169123.273\n",
      "3. Current log-likelihood: -169123.227\n",
      "4. Current log-likelihood: -169123.188\n",
      "5. Current log-likelihood: -169123.152\n",
      "6. Current log-likelihood: -169123.131\n",
      "7. Current log-likelihood: -169123.120\n",
      "8. Current log-likelihood: -169123.100\n",
      "9. Current log-likelihood: -169123.081\n",
      "10. Current log-likelihood: -169123.061\n",
      "11. Current log-likelihood: -169123.043\n",
      "12. Current log-likelihood: -169123.025\n",
      "Optimal log-likelihood: -169123.012\n",
      "Rate parameters:  A-C: 8.02668  A-G: 9.95859  A-T: 2.37802  C-G: 4.09924  C-T: 23.52959  G-T: 1.00000\n",
      "Base frequencies:  A: 0.330  C: 0.156  G: 0.211  T: 0.303\n",
      "Site proportion and rates:  (0.209,0.003) (0.139,0.043) (0.136,0.183) (0.161,0.509) (0.125,0.906) (0.138,1.898) (0.092,5.561)\n",
      "Parameters optimization took 12 rounds (3.562 sec)\n",
      "BEST SCORE FOUND : -169123.012\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file demo_data/tree_ufb.splits.nex\n",
      "Total tree length: 24.008\n",
      "\n",
      "Total number of iterations: 102\n",
      "CPU time used for tree search: 272.124 sec (0h:4m:32s)\n",
      "Wall-clock time used for tree search: 49.119 sec (0h:0m:49s)\n",
      "Total CPU time used: 704.863 sec (0h:11m:44s)\n",
      "Total wall-clock time used: 153.172 sec (0h:2m:33s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file demo_data/tree_ufb.splits.nex...\n",
      "39 taxa and 113 splits.\n",
      "Consensus tree written to demo_data/tree_ufb.contree\n",
      "Reading input trees file demo_data/tree_ufb.contree\n",
      "Log-likelihood of consensus tree: -169123.005\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                demo_data/tree_ufb.iqtree\n",
      "  Maximum-likelihood tree:       demo_data/tree_ufb.treefile\n",
      "  Likelihood distances:          demo_data/tree_ufb.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          demo_data/tree_ufb.splits.nex\n",
      "  Consensus tree:                demo_data/tree_ufb.contree\n",
      "  Screen log file:               demo_data/tree_ufb.log\n",
      "\n",
      "Date and Time: Thu Apr 25 23:24:18 2024\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -s demo_data/all_seqs_mafft.fa -m GTR+F+R7 -pre demo_data/tree_ufb -bb 1000 -nt AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **Step 5. Visualization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Please proceed to [iTOL](https://itol.embl.de/)**<br>\n",
    "**Upload `demo_data/tree_ufb.treefile` file**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dstu_hantavirus_phylo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
